tests/test/unsize.rs

```diff
@@ -586,3 +586,200 @@ fn struct_unsizing() {
         }
     }
 }
+
+#[test]
+fn dispatch_from_dyn() {
+    test! {
+        program {
+            #[lang(dispatch_from_dyn)]
+            trait DispatchFromDyn<T> {}
+
+            impl<'a, T, U> DispatchFromDyn<&'a U> for &'a T {}
+        }
+
+        // Smoke test that DispatchFromDyn works just like any other impl.
+        goal {
+            forall<'a> {
+                &'a u8: DispatchFromDyn<&'a u8>
+            }
+        } yields {
+            "Unique"
+        }
+    }
+}
+
+#[test]
+fn dispatch_from_dyn_wf() {
+    lowering_success! {
+        program {
+            #[lang(dispatch_from_dyn)]
+            trait DispatchFromDyn<T> {}
+
+            #[one_zst]
+            struct PhantomData<T> {}
+
+            struct Foo<T> {
+                f: *mut T,
+                f2: PhantomData<u8>,
+            }
+
+            // References and pointers
+            impl<'a, T, U> DispatchFromDyn<&'a U> for &'a T {}
+            impl<'a, T, U> DispatchFromDyn<&'a mut U> for &'a mut T {}
+            impl<T, U> DispatchFromDyn<*const U> for *const T {}
+            impl<T, U> DispatchFromDyn<*mut U> for *mut T {}
+
+            // Struct
+            impl<T, U> DispatchFromDyn<Foo<U>> for Foo<T> {}
```

Comment:
So I'm just not realizing that 1-ZSTs are fine. The wording above in wf.rs makes it seem like any 1-ZSTs are forbidden. I guess they're just ignored. Can you clarify that above?